### PR TITLE
Fix attr table sorting

### LINKED ISSUE #34935: "sort" funcionality of attributes table sort numeric fields as text
3.10.3 and 3.12

Add the "alaska" shapefile from the QGIS sample dataset

Open the attribute table and use the "sort" function to sort the records using the "cat" field, that is of type integer.

The records will be ordered as expected.

Save the project.

Close the project and reopen it, open the attribute table > records are ordered as the "cat" column was of type text (i.e. 1,10,100,11,110,111...).

### PR DESCRIPTION
**Pull request by elpaso**
Fix #34935

### CODE REVIEW

File: src/gui/attributetable/qgsattributetablemodel.cpp [hunk 1]
```diff
@@ -235,7 +236,13 @@ void QgsAttributeTableModel::featureAdded( QgsFeatureId fid )
...
         const QVariant &widgetCache = mAttributeWidgetCaches.at( cache.sortFieldIndex );
         const QVariantMap &widgetConfig = mWidgetConfigs.at( cache.sortFieldIndex );
-        const QVariant sortValue = fieldFormatter->representValue( mLayer, cache.sortFieldIndex, widgetConfig, widgetCache, mFeat.attribute( cache.sortFieldIndex ) );
+        // If using the default formatter use the raw value for sorting
+        // (keep the correct QVariant type and do not convert to a possibly localized string)
+        // See: https://github.com/qgis/QGIS/issues/34935
+        const bool isFallbackFormatter { fieldFormatter->id().isEmpty() };
+        const QVariant sortValue = isFallbackFormatter ?
+                                   mFeat.attribute( cache.sortFieldIndex ) :
+                                   fieldFormatter->representValue( mLayer, cache.sortFieldIndex, widgetConfig, widgetCache, mFeat.attribute( cache.sortFieldIndex ) );
```
Comment (nyalldawson) [inline src/gui/attributetable/qgsattributetablemodel.cpp:245]: I think this should be using fieldFormatter->sortValue( ... ) instead. That already has appropriate logic in the field formatter subclasses to correctly handle sort order (and is used elsewhere in this class)

File: src/gui/attributetable/qgsattributetablemodel.cpp [hunk 2]
```diff
@@ -33,6 +33,7 @@
 #include "qgsvectordataprovider.h"
 #include "qgssymbollayerutils.h"
 #include "qgsfieldformatterregistry.h"
+#include "qgsfallbackfieldformatter.h"
```
Comment (m-kuhn) [inline src/gui/attributetable/qgsattributetablemodel.cpp:36]: Is that needed?

**Review by m-kuhn — COMMENTED** [review]
LGTM

File: src/gui/attributetable/qgsattributetablemodel.cpp (hunk 2 above)
Comment (nyalldawson) [inline src/gui/attributetable/qgsattributetablemodel.cpp:36]:
```suggestion
```

**Review by nyalldawson — APPROVED** [review]

**Pull request merged**